sim-lib/src/cln.rs

```diff
@@ -96,16 +96,15 @@ impl LightningNode for ClnNode {
         dest: PublicKey,
         amount_msat: u64,
     ) -> Result<PaymentHash, LightningError> {
-        let KeysendResponse { payment_hash, .. } = self
+        let response = self
             .client
             .key_send(KeysendRequest {
                 destination: dest.serialize().to_vec(),
                 amount_msat: Some(Amount { msat: amount_msat }),
                 ..Default::default()
             })
-            .await
-            .map_err(|err| LightningError::SendPaymentError(err.to_string()))?
-            .into_inner();
+            .await?;
```

Comment:
So, this is where the different lightning implementations behavior starts to bite us. 
I believe that these two are equivalent: 
* LND API call succeeds, then `TrackPayment` gives us `RouteNotFound`. 
* CLN API call fails with error code 210 "Ran out of routes to try"

While the approach of a zero payment hash _works_, I think that it makes the meaning of a `Ok` result kind of confusing here. 

An alternative could be something like: 
* Define a `LightningError` variant `SendPaymentFailure` that has a `PaymentOutcome` associated with it. 
* Add a new `ActionOutcome` (now `SimulationOutcome` `PaymentFailed` and send it along to `produce_simulation_results` like we do for payments that we could dispatch. 
* Rather than dispatch tracking in `produce_simulation_results`, we can just pipe the result right into the channel because we already know the outcome. 

This is a lot more code, but it does take a more "standard" path for handling the results. Do you think it's worth it?

Reply:
> Add a new `ActionOutcome` (now `SimulationOutput`) PaymentFailed and send it along to `produce_simulation_results` like we do for payments that we could dispatch.

We'd need variants of `SimulationOutput` (formerly  `ActionOutcome`), like you propose, but for each `ActivityDefinition` type we add. I expect we'd have variants of `produce_simulation_results` and `consume_simulation_results` that service each `ActivityDefinition` and should confine result reporting to the channels defined by the pair.

> Rather than dispatch tracking in `produce_simulation_results`, we can just pipe the result right into the channel because we already know the outcome.

Looks like a complement to zero payment hash. Since the channel within which we need to stick the result is of type `Sender<(DispatchedPayment, PaymentResult)>`, we still need to have an optional payment hash within the sender, else we'd have to differentiate the channel in which we pipe the results. See https://github.com/bitcoin-dev-project/sim-ln/pull/104/commits/0615f4bc2f7fe4c25e16cf650ec19f61b0100dc0

> Define a `LightningError` variant `SendPaymentFailure` that has a `PaymentOutcome` associated with it.

`PaymentOutcome` is not an error and it's probably no good to carry non-error payloads within errors. Besides, we don't need it

Reply:
> Looks like a complement to zero payment hash

Yeah indeed, didn't think that one all the way through. All for the zero hash wrapped in the success/failure enum, just don't like having it as the value we based logic on.